kommandir/bin/adept_openstack.py

```diff
@@ -0,0 +1,720 @@
+        return floating_ip
+
+
+def create(name, pub_key_files):
+    """
+    Create a new VM with name and authorized_keys containing pub_key_files.
+
+    :param name: Name of the VM to create
+    :param pub_key_files: List of ssh public key files to read
+    :param basename: Not used
+    """
+
+    pubkeys = []
+    for pub_key_file in pub_key_files:
+        logging.debug("Loading public key file: %s", pub_key_file)
+        with open(pub_key_file, 'rb') as key_file:
+            pubkeys.append(key_file.read().strip())
+            if 'PRIVATE KEY' in pubkeys[-1]:
+                raise ValueError("File %s appears to be a private, not a public, key"
+                                 % pub_key_file)
+    logging.info("Deleting any partially created VM %s", name)
+    TimeoutDeleted(name)()
+
+    try:
+        server_id = TimeoutCreate(name, pubkeys)()
+        logging.info("Creation successful, attempting to assign floating ip to VM %s", name)
+        floating_ip = TimeoutAssignFloatingIP(server_id)()
+        logging.info("Floating IP assignment successful, ip %s", floating_ip)
+    except:
+        # Fire and forget
+        os_rest = OpenstackREST()
+        os_rest.server_delete(name)
+        raise
+    sys.stdout.write(OUTPUT_FORMAT.format(name=name, floating_ip=floating_ip))
+    return 0
+
+
+def discover(name, pub_key_files=None):
+    """
+    Write ansible host_vars to stdout if a VM name exists with a floating IP.
+
+    :param name: Name of the VM to search for
+    :param pub_key_files: Not used
+    """
+    # Allows earlier CLI parameter error detection
+    del pub_key_files
+
+    os_rest = OpenstackREST()
+    # Server is useless if it can't be reached
+    floating_ip = os_rest.server_ip(name)
+
+    sys.stdout.write(OUTPUT_FORMAT.format(name=name, floating_ip=floating_ip))
+    return 0
+
+
+def destroy(name):
+    """
+    Destroy VM name
+
+    :param name: Name of the VM to destroy
+    """
+    TimeoutDeleted(name)()
+    return 0
+
+
+def parse_args(argv, operation):
+    """
+    Examine command line arguments, show usage info if inappropriate for operation
+
+    :param argv: List of command-line arguments
+    :param operation: String of 'discover', 'create', or 'destroy'
+    :returns: Dictionary of parsed command-line options
+    """
+    # N/B argv[0] is path to executed command
+    argv = argv[1:]
+    try:
+        parsed_args = dict(name=argv.pop(0))
+    except IndexError:
+        raise ValueError("Must pass server name as first parameter")
+    if operation in ['create', 'discover']:
+        if len(argv) < 1:
+            raise ValueError("Must pass server name, and one or more"
+                             " paths to public ssh key files as parameters")
+        parsed_args['pub_key_files'] = set(argv)
+    elif operation == 'help':
+        logging.info("FIXME: Some useful --help message")
+    elif operation not in ('discover', 'destroy'):
```

Comment:
Why not just `elif operation != 'destroy'`? The 'discover' case is handled above.